src/anyp/Uri.cc

```diff
@@ -514,27 +515,37 @@ AnyP::Uri::absolute() const
  *        After copying it on in the first place! Would be less code to merge the two with a flag parameter.
  *        and never copy the query-string part in the first place
  */
-char *
+SBuf
 urlCanonicalCleanWithoutRequest(const SBuf &url, const HttpRequestMethod &method, const AnyP::UriScheme &scheme)
 {
-    LOCAL_ARRAY(char, buf, MAX_URL);
-
-    snprintf(buf, sizeof(buf), SQUIDSBUFPH, SQUIDSBUFPRINT(url));
-    buf[sizeof(buf)-1] = '\0';
-
     // URN, CONNECT method, and non-stripped URIs can go straight out
+    SBuf out = url;
     if (Config.onoff.strip_query_terms && !(method == Http::METHOD_CONNECT || scheme == AnyP::PROTO_URN)) {
-        // strip anything AFTER a question-mark
-        // leaving the '?' in place
-        if (auto t = strchr(buf, '?')) {
-            *(++t) = '\0';
-        }
+        // strip anything after a path segment
+        // leaving the delimiter in place as a visual cue of truncation (if any)
+        static const CharacterSet pathDelims("path-delim", "?#");
+        const auto pathEnd = url.findFirstOf(pathDelims);
+        if (pathEnd != SBuf::npos)
+            out = url.substr(0, pathEnd+1);
     }
 
-    if (stringHasCntl(buf))
-        xstrncpy(buf, rfc1738_escape_unescaped(buf), MAX_URL);
+    // check for unescaped control characters
+    // XXX: emulates old stringHasCtl() which matched true
+    //      only for the UTF-8 characters 0x80-09f.
+    // TODO: RFC 1738, 5234 and 7230 require the full UTF-8
+    //      range (0x80-0xFF) be encoded along with
+    //      the ASCII control characters 0x01-0x1F, 0x7f
+    //      and NUL (0x00)
+    static const CharacterSet partialUtf8 = CharacterSet("utf8-partial", 0x80, 0x9f);
+    static const CharacterSet ctls = (partialUtf8 + CharacterSet::CTL)
+                                     .add('\0')
+                                     .rename("controls");
+    if (out.findFirstOf(ctls) != SBuf::npos) {
+        SBuf tmp = out;
+        out = SBuf(rfc1738_escape_unescaped(tmp.c_str()));
```

Comment:
If you recall we had issues with c_str()'s memory allocation being invalidated before the function using it was actually done when X.c_str() used as a parameter and the SBuf X involved with other changes on the same line/statement. Especially for builds with high levels of compiler optimization happening (eg "-O3").

The tmp is kept in the if-statement sub-scope to minimize that period of extra memory use.

If the coder API was altered to take c-string lengths we could avoid this entirely and just use rawContent() + length(). But changes to that coder are intentionally out of scope here.

Reply:
You have committed the suggested SBuf() improvement, but this change request was focusing on the unnecessary (and harmful) copying of `out` to `tmp`. After the committed improvement, that `tmp` local became completely unused, but it still harmful (because it disabled SBuf::cow() optimizations inside c_str()). Please remove `tmp` and the corresponding copying line.

Reply:
This should now be resolved with 198201e using the SBuf based encoder.